feat: support Github Copilot

eeee0717 opened the pull request:
#2205
- [x] logout删除本地copilot token
- [x] 登录显示用户名 
- [x] 自定义headers
- [x] 连接请求时间
- [x] add rate limit
![CleanShot 2025-03-10 at 19 22 06@2x](https://github.com/user-attachments/assets/e0317d9b-40e0-4ece-bd76-129a5ce336cd)

eeee0717 committed:
feat: support Github Copilot

eeee0717 committed:
feat: finish i18n translate

eeee0717 committed:
Merge branch 'main' into feat-github-model

eeee0717 commented:
感谢ChatWise的作者@egoist提供的思路帮助

eeee0717 commented:
> 效率太高了 👏

还得感谢大佬的指点🙏

FischLu commented:
这个好, 订阅了 copilot 等于可以无限使用这几个模型了, 我测了一下下面几个都 ok
<img width="592" alt="image" src="https://github.com/user-attachments/assets/4ee0c058-9d1f-4d84-8c05-ed618d838559" />

eeee0717 commented:
> 请教一下，Cursor 的订阅用户，是否有可能也采用类似的办法接入？

这个我不知道诶，我没有cursor的订阅

kangfenmao commented:
下个版本合并，多测试测试

onevcat commented on src/main/services/CopilotService.ts:
```diff
@@ -0,0 +1,143 @@
+   */
+  public async saveCopilotToken(_: Electron.IpcMainInvokeEvent, token: string): Promise<void> {
+    try {
+      await fs.writeFile(this.TOKEN_FILE_PATH, token)
```
直接明文保存这个GitHub token是不是带来安全隐患（被别的非沙盒app直接读取）？建议是不是至少 SHA 一下或者能不能考虑 [safe storage](https://www.electronjs.org/docs/latest/api/safe-storage)?

eeee0717 replied:
ok，我看一下

onevcat replied:
感谢。不过 safe storage 的话可能会导致认证状态无法通过 WebDAV 进行备份和同步，这里需要在安全和易用间做一个权衡...

eeee0717 replied:
感觉还是安全性优先吧，毕竟如果token被盗了导致封号就不好了，安全性这块确实没考虑到，感谢您的提醒🎉

onevcat replied:
感谢修改！其实我第一天接触 Cherry Studio 的代码库，很多地方还完全不理解。不过看起来其他 LLM Provider 的 API Key 也是直接在 store/index.ts 里用 `import storage from 'redux-persist/lib/storage'` 明文保存到 Local Storage 里了..

<img width="638" alt="截屏2025-02-27 21 35 35" src="https://github.com/user-attachments/assets/6b870d95-80d4-4079-b6df-4e7976447454" />

如果 Cherry Studio 之后做大了的话，也许会成为问题：任意其他的第三方软件都可以很容易读取到这些 API Key 并且滥用，给 Cherry Studio 造成信誉上的打击。不知道主要维护者怎么看待这个问题：要不要重视这种风险（现阶段也许没什么大问题），以及如何对应（比如至少把 API Key 用 safe storage 包一下）之类的。

这方面不是我现在一个刚开始看代码的小白可以判断的了

（不是特别理解 Cherry Studio repo 的运作方式，请允许我 mention 一下 @kangfenmao 作为提醒 XD）

eeee0717 replied:
昨天和作者讨论了一下这个问题，短期内暂时不会调整，后续如果加入插件系统会修改这部分逻辑，也参考了一下OpenWebUI这个项目（https://github.com/open-webui/open-webui）他们是存了个加密token在前端，然后在后端解析。
后续应该也会采用类似的方案，如果您对这部分感兴趣，也欢迎加入我们开发群一起讨论！

kangfenmao replied:
后面找一个 security storage 改下存储位置就好了

eeee0717 committed:
fix: add safeStorage

eeee0717 committed:
Merge branch 'main' into feat-github-model

eeee0717 committed:
clean code

zxjzxj0010 commented:
参考 chatwise，只有 gpt4o 是支持图像的。
![416001362-57dcdc3e-b5d9-45ee-aa9d-51a2db53be5c.png](https://github.com/user-attachments/assets/a87fee1b-dbfc-462a-9758-5ded1e46713e)

FischLu commented:
我测的4o也不行, copilot 的没有一个识别图像的

eeee0717 committed:
chore: remove vision model

eeee0717 committed:
Merge branch 'main' into feat-github-model

eeee0717 commented:
![CleanShot 2025-03-01 at 10 31 49@2x](https://github.com/user-attachments/assets/a1f74e50-1d19-4af9-9648-6e05bf7fbba4)
已取消视觉模型标记

eeee0717 committed:
Merge branch 'main' into feat-github-model

dabing1022 commented:
期待早日 review 下个版本发布 😂

momentahjun commented:
等合并

eeee0717 committed:
Merge remote-tracking branch 'upstream/main' into feat-github-model

kangfenmao committed:
Merge branch 'main' into feat-github-model

vaayne committed:
✨ feat: add Model Context Protocol (MCP) support (#2809)

* ✨ feat: add Model Context Protocol (MCP) server configuration (main)

- Added `@modelcontextprotocol/sdk` dependency for MCP integration.
- Introduced MCP server configuration UI in settings with add, edit, delete, and activation functionalities.
- Created `useMCPServers` hook to manage MCP server state and actions.
- Added i18n support for MCP settings with translation keys.
- Integrated MCP settings into the application's settings navigation and routing.
- Implemented Redux state management for MCP servers.
- Updated `yarn.lock` with new dependencies and their resolutions.

* 🌟 feat: implement mcp service and integrate with ipc handlers

- Added `MCPService` class to manage Model Context Protocol servers.
- Implemented various handlers in `ipc.ts` for managing MCP servers including listing, adding, updating, deleting, and activating/deactivating servers.
- Integrated MCP related types into existing type declarations for consistency across the application.
- Updated `preload` to expose new MCP related APIs to the renderer process.
- Enhanced `MCPSettings` component to interact directly with the new MCP service for adding, updating, deleting servers and setting their active states.
- Introduced selectors in the MCP Redux slice for fetching active and all servers from the store.
- Moved MCP types to a centralized location in `@renderer/types` for reuse across different parts of the application.

* feat: enhance MCPService initialization to prevent recursive calls and improve error handling

* feat: enhance MCP integration by adding MCPTool type and updating related methods

* feat: implement streaming support for tool calls in OpenAIProvider and enhance message processing

zhweji commented:
期待

kangfenmao committed:
fix: finish_reason undefined

dabing1022 commented:
能否尽快合并？

eeee0717 committed:
Merge remote-tracking branch 'upstream/main' into feat-github-model

eeee0717 committed:
fix migrate

beyondkmp commented:
期待

beyondkmp commented:
可以直接自己编译一个，下载https://github.com/eeee0717/cherry-studio/tree/feat-github-model,然后编译安装下。

eeee0717 committed:
Merge remote-tracking branch 'upstream/main' into feat-github-model

duckkkey commented:
请教一个关于copilot账号的问题：对Copilot的异常使用，尤其是多次触发429，可能导致封号，比如：
https://github.com/jjleng/copilot-more/issues/50
https://github.com/cline/cline/issues/1972#issuecomment-2692872820
请问是否考虑增加提醒，或者加入避免封号的措施，比如速率限制呢？

eeee0717 commented:
> 请教一个关于copilot账号的问题：对Copilot的异常使用，尤其是多次触发429，可能导致封号，比如： [jjleng/copilot-more#50](https://github.com/jjleng/copilot-more/issues/50) [cline/cline#1972 (comment)](https://github.com/cline/cline/issues/1972#issuecomment-2692872820) 请问是否考虑增加提醒，或者加入避免封号的措施，比如速率限制呢？

之前有考虑过合规性的问题，由于我本身没有遇到过429的情况，所以就忽视了，我会考虑加上提醒和速率限制的，感谢您的提醒！

duckkkey commented:
> > 请教一个关于copilot账号的问题：对Copilot的异常使用，尤其是多次触发429，可能导致封号，比如： [jjleng/copilot-more#50](https://github.com/jjleng/copilot-more/issues/50) [cline/cline#1972 (comment)](https://github.com/cline/cline/issues/1972#issuecomment-2692872820) 请问是否考虑增加提醒，或者加入避免封号的措施，比如速率限制呢？
> 
> 之前有考虑过合规性的问题，由于我本身没有遇到过429的情况，所以就忽视了，我会考虑加上提醒和速率限制的，感谢您的提醒！

不客气，期待尽快合并！👍

eeee0717 committed:
feat: add rate limit and warning

eeee0717 commented:
由于目前消息模块在重构，暂时无法合并速率限制相关代码
![CleanShot 2025-03-09 at 11 04 45@2x](https://github.com/user-attachments/assets/d59be062-5b6f-47dd-9651-a2fb53b92836)

beyondkmp commented:
![image](https://github.com/user-attachments/assets/7e194798-e968-435f-b04a-73becfa4ade3)
我有两个github账号，有时候切换登录 ，浏览器已经显示成功了，cherry studio一直没有成功，会不会是1s发一次，请求太多了？一直成功不了，需要过一段时间手动再点下`连接github`. 想问下是一定要再次手动点吗？还是会自动连接上？

eeee0717 commented:
需要再次手动点

beyondkmp commented on src/main/services/CopilotService.ts:
```diff
@@ -0,0 +1,144 @@
+    const pollingInterval = 1000 // 1秒
+
+    for (let attempts = 0; attempts < maxAttempts; attempts++) {
+      await this.delay(pollingInterval)
```
这里只间隔1s, 会不会太快太多？如果真失败了是不是用`指数退避重试`会更好？

beyondkmp commented on src/main/services/CopilotService.ts:
```diff
@@ -0,0 +1,144 @@
+  constructor() {
+    this.GITHUB_CLIENT_ID = 'Iv1.b507a08c87ecfe98'
+    this.TOKEN_FILE_PATH = path.join(app.getPath('userData'), '.copilot_token')
+    this.DEFAULT_HEADERS = {
```
这个default headers，可以在配置自定义吗？这样可以有更多的选择，比如用zed的agent，懂的人可以自定义。或者有选项的也行，这样可以分开多个不同的agent header, github 服务端会更加难的来block请求。

beyondkmp commented:
还有一个小小的需求，我有两个github账号，现在这样显示不知道是哪个登录成功了，可以显示下在这个设置页面显示出登录用户的用户名吗？
![image](https://github.com/user-attachments/assets/4259f575-ca59-4132-861c-80e40be79aa1)

eeee0717 commented:
> 还有一个小小的需求，我有两个github账号，现在这样显示不知道是哪个登录的，可以在这个设置页面显示出登录用户的用户名吗？ ![image](https://private-user-images.githubusercontent.com/5462444/420665507-4259f575-ca59-4132-861c-80e40be79aa1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE1MTg0ODUsIm5iZiI6MTc0MTUxODE4NSwicGF0aCI6Ii81NDYyNDQ0LzQyMDY2NTUwNy00MjU5ZjU3NS1jYTU5LTQxMzItODYxYy04MGU0MGJlNzlhYTEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMDlUMTEwMzA1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2E2YjkyYWRiMmMxYzU2NzY4YzY2ZDMxMWVmZWM1Zjg3NmIxMzM0OTYzYjc0OTI1NzBiYTEzNDM0M2EwYTVmNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.EfDAIdXEPKPMIZhVjdhiwtDIhwyQZTg-rW72oJNJews)

OK 上述反馈均已收到，明天我会考虑修改

eeee0717 commented:
@beyondkmp 那是不是还可以加一个多账号登录的功能

beyondkmp commented:
不过现在可以添加两个github copilot的模型，也算是多账号登录了。如果直接在一个页面就有多账号登录，那就更加完美了，两个账号的模型不太一样，有的是enable了claude3.7, 有的没有, 所以有这个功能后就可以互补了。现在zed/vim/vscode都不支持多账号同时登录 ，有一点风险是要注意的：会不会有其它安全合规的问题或者被github 服务端检测到异常？

eeee0717 commented:
目前来看只有上面提到的速率问题可能会引发风险，暂时还没了解到其他的情况，我自己在vscode/nvim上都没有触发过风险问题，只能是再三提醒用户可能会存在风险。还是先把上面的一些问题弄好再后续考虑多账号的情况。

beyondkmp commented:
还有一个问题，退出登录后，这个.copilot_token文件没有删除掉，这样感觉不太安全。
![image](https://github.com/user-attachments/assets/c528d5d2-836a-4467-89b5-4f232e1eb937)

eeee0717 commented:
ok

beyondkmp commented on src/renderer/src/pages/settings/ProviderSettings/GithubCopilotSettings.tsx:
```diff
@@ -0,0 +1,225 @@
+    if (provider.isAuthed) {
+      setAuthStatus(AuthStatus.AUTHENTICATED)
+    } else {
+      setAuthStatus(AuthStatus.NOT_STARTED)
```
logout后是不是也需要向github发一个请求？我现在的问题是两个账号，登出后再用另外一个账号登录，登录后还是前一个账号。

ViGeng commented:
弱弱的讨论一下，这样玩会不会把Github绕闹了，给封掉。Cherry Studio也算个很有影响力的社区了，感觉可以考虑一下这种特性是否合适。当然我要给PR作者点个赞👍

FischLu commented:
> 弱弱的讨论一下，这样玩会不会把Github绕闹了，给封掉。Cherry Studio也算个很有影响力的社区了，感觉可以考虑一下这种特性是否合适。当然我要给PR作者点个赞👍

GitHub 为什么要封, 他自己也提供 copilot 网页版了, 这个 PR 无非就是把 user interface 挪到客户端了, 而且也不是白嫖, 不订阅也用不了

eeee0717 commented:
> 弱弱的讨论一下，这样玩会不会把Github绕闹了，给封掉。Cherry Studio也算个很有影响力的社区了，感觉可以考虑一下这种特性是否合适。当然我要给PR作者点个赞👍

不少客户端都通过这样的方式使用copilot，chatwise，vscode，vim，zed等，目前来看只要注意速率限制就可以的，而且已经添加了风险提醒的二次确认，如果觉得有风险也可以不使用。

eeee0717 commented:
- [x] logout删除本地copilot token
- [x] 登录显示用户名 
- [x] 自定义headers
![CleanShot 2025-03-10 at 19 22 06@2x](https://github.com/user-attachments/assets/e0317d9b-40e0-4ece-bd76-129a5ce336cd)
- [ ] 连接请求时间

eeee0717 committed:
feat: add delete copilot token file

feat: add login message

feat: add default headers and change getCopilotToken algorithm

eeee0717 committed:
fix

eeee0717 commented:
应该是migrate这里的问题，直接push了
![CleanShot 2025-03-10 at 20 28 51@2x](https://github.com/user-attachments/assets/ca2ceb87-fa6a-4efb-8ab2-5fc3f1bee93f)

beyondkmp commented:
好的，多谢，我测试了下没有复现。

你这个分支的右键复制有些问题，有空帮忙merge下最新的代码，非常感谢。

![image](https://github.com/user-attachments/assets/d3b459e3-3e3c-44f0-8d85-f2ee772a7e87)

eeee0717 commented:
我还在等message部分重构完成，等完成后我会尽快merge，并且加上rate limit

beyondkmp commented:
希望能 merge 下，这样可以
我可以自己再 build 一个版本，多测试下😁

eeee0717 committed:
Merge remote-tracking branch 'upstream/main' into feat-github-model

eeee0717 committed:
feat: add rate limit

eeee0717 committed:
Merge branch 'main' into feat-github-model

eeee0717 commented:
- [x] add rate limit

eeee0717 committed:
Merge remote-tracking branch 'upstream/main' into feat-github-model

dabing1022 commented:
Oh my goodness! We are really hoping for a review and release as soon as possible! It has truly been such a long time and we simply can't wait any longer! 😊

eeee0717 commented:
> @eeee0717 多谢merge最新代码。用起来很爽了，也没有发现新的问题了，相比于vscode里面的那个chat窗口方便多了。太感谢了🙏!

我测试下来使用copilot claude调用mcp会稳定报错，后面还需要解决一下。

eeee0717 committed:
chore: change apihost

beyondkmp commented:
> > @eeee0717 多谢merge最新代码。用起来很爽了，也没有发现新的问题了，相比于vscode里面的那个chat窗口方便多了。太感谢了🙏!
> 
> 我测试下来使用copilot claude调用mcp会稳定报错，后面还需要解决一下。

了解，多谢提醒，还没用这个mcp功能。这个是不是后面再建PR解决更好？

eeee0717 commented:
对的 我只是记录提醒一下发现的问题，怕后面忘记了😭

FischLu commented on src/renderer/src/pages/settings/ProviderSettings/ProviderSetting.tsx:
```diff
@@ -244,9 +245,12 @@ const ProviderSetting: FC<Props> = ({ provider: _provider }) => {
     setApiKey(provider.apiKey)
     setApiHost(provider.apiHost)
-  }, [provider.apiKey, provider.apiHost])
+  }, [provider.apiKey, provider.apiHost, provider.id, apiKey])
```
这里不需要监听最后一个 apiKey, 函数体中也没有使用, 监听了 apiKey 会导致其他模型供应商输入的 api Key 无法删除

eeee0717 committed:
fix: remove duplicate apikey

kangfenmao commented:
mcp 稳定了之后合并

beyondkmp commented on src/renderer/src/store/llm.ts:
```diff
@@ -187,6 +187,17 @@ const initialState: LlmState = {
+      name: 'Github Copilot',
+      type: 'openai',
+      apiKey: '',
+      apiHost: 'https://api.individual.githubcopilot.com/',
```
使用这个域名，会获取不到模型。之前的api.githubcopilot.com就没有问题。

eeee0717 replied:
我看vscode用的这个，可能和headers有关系，我改回去了

eeee0717 committed:
fix: change api host

beyondkmp commented:
最近使用还发现两个问题
1. 如果用户在点登录，然后用户没有点开device连接打开github页面，而是再次去点连接github按键，这个时候一直转，应该是去发请求看有没有授权了，但是最后失败了，然后再点device连接打开github页面授权，授权成功后再来cherry studio上面点连接github，就连接不成功了。
2. 先登录一github账号，没有claude3.7模型，获取了一次模型，就存下来了，再退出登录这个账号，登录另外一个有claude3.7的账号，选模型的时候一直不会出来claude3.7, 只有全部清除缓存和重置后，再登录这个有claude3.7的账号，选模型才会有3.7出来了，目前不太清楚是github的问题，还是cherry studio的问题。

上面的问题可能一般用户不会遇到，不过用户基数大了，就有可能会有人遇到。有空的时候可以帮忙看下。

eeee0717 committed:
chore: add vertify first tooltip

eeee0717 commented:
第一个问题修复了，第二个我手里没有两个github账号，先放放吧，如果之后有别人也遇到这个问题再修改

eeee0717 committed:
Merge branch 'main' into feat-github-model

mmonaco14 commented:
> ![CleanShot 2025-03-01 at 10 10 12@2x](https://private-user-images.githubusercontent.com/70054568/418220799-a81ec381-b98f-45bf-8791-ecd6d0282ce3.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI0NTEyNTksIm5iZiI6MTc0MjQ1MDk1OSwicGF0aCI6Ii83MDA1NDU2OC80MTgyMjA3OTktYTgxZWMzODEtYjk4Zi00NWJmLTg3OTEtZWNkNmQwMjgyY2UzLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzIwVDA2MDkxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTczNDY5NWYyZTVkOTMxOTRiMzdhZThhMmU2MGMzYzUxOWY0Y2U3MDUzNTdmZGRkNGEyNzk3ZmI1NGI5NDhiN2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.lIyKeGkmV1YVm3zBwPOgzMPtwhwyVwTt_ItrhNRbZOc) 我在vscode里测的4o以及其他的也都不能识别图像 不知道chatwise是怎么实现的

可以看一下 copilot 的文档   官方说之后 4o 支持  
![局部截取_20250320_135616](https://github.com/user-attachments/assets/2277d7f3-e461-4024-a6bc-0752adae026a)

eeee0717 commented:
> 1.1.8 版本，登录会显示报错：
> 
> ![image](https://private-user-images.githubusercontent.com/16911282/424950390-7bad3790-d2c4-4d10-9fe4-a3ec22976312.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI0NzA2NDQsIm5iZiI6MTc0MjQ3MDM0NCwicGF0aCI6Ii8xNjkxMTI4Mi80MjQ5NTAzOTAtN2JhZDM3OTAtZDJjNC00ZDEwLTlmZTQtYTNlYzIyOTc2MzEyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzIwVDExMzIyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI0NTZhNWMxMDMyYzgzNjUxOTc3YTY0ZGRiZDc1Y2ZlZjcwMDhjMTRmM2E2M2FkODJiMDBkNGU0OTllNzBjZTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.YRIuSC932JvrxYuUIQ61OjNHa9at7QrVWs0fsL7Q95g)

我这里可以正常登录，是网络问题吗？可以看一下控制台报什么错吗？

Nick-Hopps commented:
好像就是我代理的问题，调试请求的时候我改成 fiddler 代理就没问题了，默认是走了我美国的梯子

eeee0717 commented:
> > 感谢ChatWise的作者@egoist提供的思路帮助
> 
> 
> 
> 
> 
> 
> 
> > #2205
> 
> > 
> 
> > * [x]  logout删除本地copilot token
> 
> > * [x]  登录显示用户名
> 
> > * [x]  自定义headers
> 
> > * [x]  连接请求时间
> 
> > * [x]  add rate limit
> 
> >   ![CleanShot 2025-03-10 at 19 22 06@2x](https://private-user-images.githubusercontent.com/70054568/420885599-e0317d9b-40e0-4ece-bd76-129a5ce336cd.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI0NzQ2MTMsIm5iZiI6MTc0MjQ3NDMxMywicGF0aCI6Ii83MDA1NDU2OC80MjA4ODU1OTktZTAzMTdkOWItNDBlMC00ZWNlLWJkNzYtMTI5YTVjZTMzNmNkLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzIwVDEyMzgzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU5Njk0NzliYzkzYmE1MWMwYWY0NTVmZjViMWVmMzg3NmY2YWFkODM3NTFjMGVlZDdlODRiZDI2ZTE0M2I1YWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.SGMb-UPxjbdgu6w76Oj8-q97hXgOzmWiyf4KzVXqQHM)
> 
> 
> 
> 请问下我在1.1.8版本，添加了请求头，但是在network看没变化，是没生效吗

可以贴一下network吗，请求是在后端发出去的，前端控制台理论上是看不到的

wangbanshan commented:
> > > 感谢ChatWise的作者@egoist提供的思路帮助
> > 
> > 
> > > #2205
> > 
> > 
> > > 
> > 
> > 
> > > * [x]   logout删除本地copilot token
> > 
> > 
> > > * [x]   登录显示用户名
> > 
> > 
> > > * [x]   自定义headers
> > 
> > 
> > > * [x]   连接请求时间
> > 
> > 
> > > * [x]   add rate limit
> > 
> > 
> > > ![CleanShot 2025-03-10 at 19 22 06@2x](https://private-user-images.githubusercontent.com/70054568/420885599-e0317d9b-40e0-4ece-bd76-129a5ce336cd.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI0NzQ2MTMsIm5iZiI6MTc0MjQ3NDMxMywicGF0aCI6Ii83MDA1NDU2OC80MjA4ODU1OTktZTAzMTdkOWItNDBlMC00ZWNlLWJkNzYtMTI5YTVjZTMzNmNkLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzIwVDEyMzgzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU5Njk0NzliYzkzYmE1MWMwYWY0NTVmZjViMWVmMzg3NmY2YWFkODM3NTFjMGVlZDdlODRiZDI2ZTE0M2I1YWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.SGMb-UPxjbdgu6w76Oj8-q97hXgOzmWiyf4KzVXqQHM)
> > 
> > 
> > 请问下我在1.1.8版本，添加了请求头，但是在network看没变化，是没生效吗
> 
> 可以贴一下network吗，请求是在后端发出去的，前端控制台理论上是看不到的

![image](https://github.com/user-attachments/assets/cc2813d0-da6f-4376-bb32-25e8a3e872ba)
![image](https://github.com/user-attachments/assets/2d445928-3311-477d-a77c-91eae10899eb)
我看的是前端的network，那可能不是实际的，我看看代码，感谢大佬
